File: hushh-adk-agents/._kushal_profile_data/Interview-Prep/Problems_By_Pattern/Dynamic_Programming/Climbing_Stairs/climbing_stairs.cs
public class Solution {
    public int ClimbStairs(int n) {
        if (n <= 2) return n;
        
        int prev2 = 1;  // F(1)
        int prev1 = 2;  // F(2)
        
        for (int i = 3; i <= n; i++) {
            int current = prev1 + prev2;
            prev2 = prev1;
            prev1 = current;
        }
        
        return prev1;
    }
}

File: hushh-adk-agents/._kushal_profile_data/Interview-Prep/Problems_By_Pattern/Dynamic_Programming/Climbing_Stairs/climbing_stairs.js
/**
 * @param {number} n
 * @return {number}
 *
 * APPROACH: Dynamic Programming (Fibonacci Pattern)
 * - To reach step n, you came from step n-1 or n-2
 * - ways(n) = ways(n-1) + ways(n-2)
 * - Use two variables for O(1) space optimization
 * - Time: O(N)
 * - Space: O(1)
 */
var climbStairs = function (n) {
  // Base cases
  if (n <= 2) return n;

  // Initialize for n=1 and n=2
  let prev2 = 1; // ways to reach step 1
  let prev1 = 2; // ways to reach step 2

  // Build up from step 3 to n
  for (let i = 3; i <= n; i++) {
    let current = prev1 + prev2; // Fibonacci: F(i) = F(i-1) + F(i-2)
    prev2 = prev1;
    prev1 = current;
  }

  return prev1;
};

// Example usage:
// console.log(climbStairs(2)); // 2
// console.log(climbStairs(3)); // 3
// console.log(climbStairs(5)); // 8
